File: scripts/check_site.py
"""Check local HTML/CSS/module references without downloading dependencies."""
from html.parser import HTMLParser
from pathlib import Path
from urllib.parse import urlsplit, unquote
import json
import re
import sys

ROOT = Path(__file__).resolve().parents[1]
errors = []
checked = 0

def check(source, value):
    global checked
    url = urlsplit(value)
    if url.scheme or url.netloc or not url.path or '${' in value:
        return
    target = ROOT / unquote(url.path.lstrip('/')) if url.path.startswith('/') else source.parent / unquote(url.path)
    checked += 1
    if not target.exists():
        errors.append(f'{source.relative_to(ROOT)}: missing {value}')

class Page(HTMLParser):
    def __init__(self, source):
        super().__init__()
        self.source = source
        self.ids = set()
        self.scripts = []
    def handle_starttag(self, tag, attrs):
        attrs = dict(attrs)
        if attrs.get('id'):
            if attrs['id'] in self.ids:
                errors.append(f'{self.source.relative_to(ROOT)}: duplicate id {attrs["id"]}')
            self.ids.add(attrs['id'])
        for key in ('src', 'href', 'poster'):
            if attrs.get(key):
                check(self.source, attrs[key])
        if tag == 'script' and attrs.get('src'):
            self.scripts.append(Path(attrs['src']).name)

pages = [ROOT/'index.html', *sorted((ROOT/'pages').glob('*.html'))]
for source in pages:
    page = Page(source)
    page.feed(source.read_text(encoding='utf-8'))
    for data, game in [('math-rpg-pools.js', 'math-rpg.js'), ('word-sort-pools.js', 'word-sort.js')]:
        if game in page.scripts and (data not in page.scripts or page.scripts.index(data) > page.scripts.index(game)):
            errors.append(f'{source.name}: {data} must load before {game}')
for source in (ROOT/'assets/css').glob('*.css'):
    for value in re.findall(r'url\(\s*[\'"]?([^\'"\)]+)', source.read_text(encoding='utf-8')):
        check(source, value.strip())
for source in (ROOT/'assets/js').rglob('*.js'):
    text = source.read_text(encoding='utf-8')
    for value in re.findall(r'(?:\bfrom\s*|\bimport\s*\(?\s*)[\'"](\.[^\'"\n]+)[\'"]', text):
        check(source, value)
config = json.loads((ROOT/'config.json').read_text(encoding='utf-8'))
assert re.fullmatch(r'\d+\.\d+\.\d+', config['version']), 'Invalid version'
assert re.fullmatch(r'\d{4}-\d{2}-\d{2}', config['lastUpdated']), 'Invalid date'
for error in errors:
    print(error)
print(f'{len(pages)} pages, {checked} local references, {len(errors)} errors')
sys.exit(bool(errors))

File: scripts/check-site.mjs
import assert from 'node:assert/strict';
import { existsSync, readFileSync, readdirSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { JSDOM } from 'jsdom';

const root = fileURLToPath(new URL('../', import.meta.url));
const read = file => readFileSync(path.join(root, file), 'utf8');
const walk = dir => readdirSync(path.join(root, dir), { withFileTypes: true }).flatMap(entry => {
  const name = `${dir}/${entry.name}`;
  return entry.isDirectory() ? walk(name) : [name];
});
const errors = [];
let checked = 0;
function check(source, url) {
  if (!url || /^(?:[a-z][\w+.-]*:|\/\/|#)/i.test(url) || url.includes('${')) return;
  const pathname = decodeURIComponent(url.split(/[?#]/)[0]);
  if (!pathname) return;
  const target = pathname.startsWith('/') ? path.join(root, pathname.slice(1)) : path.resolve(root, path.dirname(source), pathname);
  checked++;
  if (!existsSync(target)) errors.push(`${source}: missing ${url}`);
}

const pages = ['index.html', ...walk('pages').filter(file => file.endsWith('.html'))];
for (const file of pages) {
  const dom = new JSDOM(read(file));
  const document = dom.window.document;
  const ids = new Set();
  for (const node of document.querySelectorAll('[id]')) {
    if (ids.has(node.id)) errors.push(`${file}: duplicate id ${node.id}`);
    ids.add(node.id);
  }
  if (!ids.has('root')) errors.push(`${file}: missing React root`);
  for (const node of document.querySelectorAll('[src],[href],[poster]')) {
    for (const attribute of ['src', 'href', 'poster']) check(file, node.getAttribute(attribute));
  }
  dom.window.close();
}
for (const file of [...walk('assets/css'), ...walk('src')].filter(file => file.endsWith('.css'))) {
  for (const [, url] of read(file).matchAll(/url\(\s*['"]?([^'"\)]+)/g)) check(file, url.trim());
}
for (const file of walk('assets/js').filter(file => file.endsWith('.js'))) {
  for (const [, url] of read(file).matchAll(/(?:\bfrom\s*|\bimport\s*\(?\s*)['"](\.[^'"\n]+)['"]/g)) check(file, url);
}
const docs = [...readdirSync(root).filter(file => file.endsWith('.md')), ...walk('docs'), ...walk('assets/vendor'), ...walk('assets/images')].filter(file => file.endsWith('.md'));
for (const file of docs) {
  for (const [, url] of read(file).matchAll(/\]\(([^\s)]+)(?:\s+"[^"]*")?\)/g)) check(file, url);
}
const config = JSON.parse(read('config.json'));
assert.match(config.version, /^\d+\.\d+\.\d+$/);
errors.forEach(error => console.error(error));
console.log(`${pages.length} pages, ${docs.length} documents, ${checked} references, ${errors.length} errors`);
process.exitCode = errors.length ? 1 : 0;
